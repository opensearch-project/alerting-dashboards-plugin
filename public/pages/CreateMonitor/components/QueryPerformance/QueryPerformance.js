/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { Fragment } from 'react';
import _ from 'lodash';
import {
  EuiButton,
  EuiCallOut,
  EuiFlexGroup,
  EuiFlexItem,
  EuiIcon,
  EuiLink,
  EuiModal,
  EuiModalBody,
  EuiModalFooter,
  EuiModalHeader,
  EuiModalHeaderTitle,
  EuiSpacer,
  EuiText,
} from '@elastic/eui';
import { DEFAULT_EMPTY_DATA, MONITOR_TYPE } from '../../../../utils/constants';
import ContentPanel from '../../../../components/ContentPanel';

export const RECOMMENDED_DURATION = 100;
export const SEARCH_DOCUMENTATION = 'https://opensearch.org/docs/latest/search-plugins/';

const getPerformanceCallOut = () => (
  <EuiCallOut
    title={'Large queries may impact monitor and remote cluster performance'}
    color={'warning'}
    iconType={'alert'}
  >
    We recommend reducing your query size and time range or changing data sources to optimize for
    monitor performance.{' '}
    <EuiLink external href={SEARCH_DOCUMENTATION} target={'_blank'}>
      Learn more
    </EuiLink>
  </EuiCallOut>
);

export const getPerformanceModal = ({ edit, onClose, onSubmit, values }) => {
  const monitorType = _.get(
    values,
    'monitor_type',
    _.get(values, 'workflow_type', MONITOR_TYPE.QUERY_LEVEL)
  );

  let hasRemoteClusters;
  switch (monitorType) {
    case MONITOR_TYPE.CLUSTER_METRICS:
      hasRemoteClusters = !_.isEmpty(_.get(values, 'uri.clusters', []));
      break;
    default:
      // Indexes for remote clusters will store the index name in
      // the 'value' attribute of the object, not the 'label' attribute.
      hasRemoteClusters = _.get(values, 'index', [])
        .map(({ label, value }) => value || label)
        .some((indexName) => indexName.includes(':'));
  }

  return (
    <EuiModal onClose={onClose}>
      <EuiModalHeader>
        <EuiModalHeaderTitle>
          <h1>Monitor is not optimized</h1>
        </EuiModalHeaderTitle>
      </EuiModalHeader>

      <EuiModalBody>
        <EuiText>
          <p>The following use cases may impact this monitor's performance.</p>
          <ul>
            {hasRemoteClusters && <li>One or more remote indexes may affect monitor accuracy</li>}
            <li>Large queries may impact monitor and remote cluster performance</li>
          </ul>
        </EuiText>
      </EuiModalBody>

      <EuiModalFooter>
        <EuiButton fill={false} onClick={onSubmit}>
          {edit ? 'Update' : 'Create'} anyway
        </EuiButton>

        <EuiButton fill={true} onClick={onClose}>
          Reconfigure
        </EuiButton>
      </EuiModalFooter>
    </EuiModal>
  );
};

const QueryPerformance = ({ response, actions }) => {
  const monitorDuration = _.get(response, 'took', DEFAULT_EMPTY_DATA);
  const monitorDurationCallout = monitorDuration >= RECOMMENDED_DURATION;

  // TODO: Need to confirm the purpose of requestDuration.
  //  There's no explanation for it in the frontend code even back to opendistro implementation.
  const requestDuration = _.get(response, 'invalid.path', DEFAULT_EMPTY_DATA);
  const requestDurationCallout = requestDuration >= RECOMMENDED_DURATION;
  const displayPerfCallOut = monitorDurationCallout || requestDurationCallout;
  const alertIcon = (
    <>
      &nbsp;
      <EuiIcon type={'alert'} />
    </>
  );
  return (
    <Fragment>
      {displayPerfCallOut && (
        <>
          {getPerformanceCallOut()}
          <EuiSpacer />
        </>
      )}

      <ContentPanel
        title="Monitor performance"
        titleSize="s"
        panelStyles={{ paddingLeft: '10px', paddingRight: '10px' }}
        actions={actions}
      >
        <EuiSpacer size="s" />
        <EuiFlexGroup alignItems="flexStart" gutterSize="xl">
          <EuiFlexItem grow={false}>
            <EuiText size={'xs'} color={monitorDurationCallout ? 'danger' : undefined}>
              <strong>Monitor duration</strong>
              <span style={{ display: 'block' }}>
                {`${monitorDuration} ms`}
                {monitorDurationCallout ? alertIcon : undefined}
              </span>
            </EuiText>
          </EuiFlexItem>
          <EuiFlexItem grow={false}>
            <EuiText size={'xs'} color={requestDurationCallout ? 'danger' : undefined}>
              <strong>Request duration</strong>
              <span style={{ display: 'block' }}>
                {requestDuration}
                {requestDurationCallout ? alertIcon : undefined}
              </span>
            </EuiText>
          </EuiFlexItem>
          <EuiFlexItem grow={false}>
            <EuiText size="xs">
              <strong>Hits</strong>
              <span style={{ display: 'block' }}>
                {_.get(response, 'hits.total.value', DEFAULT_EMPTY_DATA)}
              </span>
            </EuiText>
          </EuiFlexItem>
        </EuiFlexGroup>
      </ContentPanel>
    </Fragment>
  );
};

export default QueryPerformance;
