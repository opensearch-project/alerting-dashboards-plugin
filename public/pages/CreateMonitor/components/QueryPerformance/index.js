/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import QueryPerformance from './QueryPerformance';

export default QueryPerformance;
