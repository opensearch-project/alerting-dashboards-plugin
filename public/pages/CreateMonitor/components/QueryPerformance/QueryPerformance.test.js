/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { render } from 'enzyme';

import QueryPerformance from './QueryPerformance';

describe('QueryPerformance', () => {
  test('renders', () => {
    const component = (
      <QueryPerformance response={{ took: 5, hits: { total: { value: 15, relation: 'eq' } } }} />
    );

    expect(render(component)).toMatchSnapshot();
  });
});
