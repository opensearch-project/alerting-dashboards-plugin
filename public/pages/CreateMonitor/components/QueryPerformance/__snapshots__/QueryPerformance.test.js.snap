// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueryPerformance renders 1`] = `
<div
  class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
  style="padding-left:10px;padding-right:10px"
>
  <div
    class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
    style="padding:0px 10px"
  >
    <div
      class="euiFlexItem"
    >
      <h3
        class="euiTitle euiTitle--small"
      >
        Monitor performance
      </h3>
    </div>
    <div
      class="euiFlexItem euiFlexItem--flexGrowZero"
    >
      <div
        class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
      >
        <div
          class="euiFlexItem"
        />
      </div>
    </div>
  </div>
  <div
    class="euiText euiText--extraSmall"
    style="padding:0px 10px"
  >
    <div
      class="euiTextColor euiTextColor--subdued"
    />
  </div>
  <hr
    class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginXSmall"
  />
  <div
    style="padding:0px 10px"
  >
    <div
      class="euiSpacer euiSpacer--s"
    />
    <div
      class="euiFlexGroup euiFlexGroup--gutterExtraLarge euiFlexGroup--alignItemsFlexStart euiFlexGroup--directionRow euiFlexGroup--responsive"
    >
      <div
        class="euiFlexItem euiFlexItem--flexGrowZero"
      >
        <div
          class="euiText euiText--extraSmall"
        >
          <strong>
            Monitor duration
          </strong>
          <span
            style="display:block"
          >
            5 ms
          </span>
        </div>
      </div>
      <div
        class="euiFlexItem euiFlexItem--flexGrowZero"
      >
        <div
          class="euiText euiText--extraSmall"
        >
          <strong>
            Request duration
          </strong>
          <span
            style="display:block"
          >
            -
          </span>
        </div>
      </div>
      <div
        class="euiFlexItem euiFlexItem--flexGrowZero"
      >
        <div
          class="euiText euiText--extraSmall"
        >
          <strong>
            Hits
          </strong>
          <span
            style="display:block"
          >
            15
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
